### PR TITLE
fix #737: Implement DangerousCompletableFutureUsage errorprone check

### LINKED ISSUE #737: CompletableFuture *async methods should require an executor
## What happened?
The overload which doesn't take an executor uses the fork join pool. Contention from misbehaving blocking work blocks execution in ways that we don't expect. The fork-join-pool doesn't report metrics we can use to diagnost problems.

## What did you want to happen?
We should require an executor to be passed to all methods which can take one. `supplyAsync`, `runAsync`, `thenRunAsync`, etc. In the unlikely case we do really want to use the fork join pool, we can pass `ForkJoinPool.commonPool()`.

### PR DESCRIPTION
**Pull request by carterkozak**
## Before this PR
We could easily accidentally enqueue work onto the ForkJoinPool.
See #737

## After this PR
==COMMIT_MSG==
Implement DangerousCompletableFutureUsage errorprone check
==COMMIT_MSG==

### CODE REVIEW

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            + "and specify the desired parallelism (i.e. the max number of concurrent tasks that will be submitted).\n"
+            + "By default, CompletableFuture uses the globally shared ForkJoinPool. Fork/join pools implement "
+            + "work-stealing, where any thread might steal a task from a different thread's queue when blocked "
+            + "waiting for a subtask to complete. This might not seem like an issue at first glance, but if you rely"
+            + "on the ForkJoinPool for short tasks extensively throughout your codebase and later on you add one "
+            + "piece of code that uses the pool for long (e.g. I/O) tasks, the other parts of your codebase that"
+            + "you'd expect to have consistent performance might experience performance degradation for no apparent "
+            + "reason.\n"
+            + "If you're absolutely certain that the ForkJoinPool is correct, please pass ForkJoinPool.commonPool()"
+            + "directly, ideally with a comment explaining why it is ideal.";
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java:62]: Slightly modified message from `DangerousParallelStreamUsage.ERROR_MESSAGE`

Comment (CRogers) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java:62]: This is great, love the explanation.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            return buildDescription(tree).setMessage(ERROR_MESSAGE).build();
+        }
+        return Description.NO_MATCH;
+    }
+
+    private static boolean isCompletionStageAsyncMethodWithoutExecutor(MethodInvocationTree tree, VisitorState state) {
+        if (COMPLETION_STAGE_ASYNC_INVOCATION.matches(tree, state)) {
+            Symbol.MethodSymbol symbol = ASTHelpers.getSymbol(tree);
+            // Should only happen if there are errors in the AST
+            if (symbol != null) {
```
Comment (CRogers) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java]: Perhaps extract the comment to a variable (ie make a var that is `boolean noErrorsInAst = symbol != null`)?

Comment (CRogers) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java]: Maybe also invert the `if` so it returns early to reduce nesting?

**Review by CRogers — APPROVED** [review]

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java (hunk 2 above)
Comment (carterkozak) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java]: I can't make that flow very well, I think the comment inside the (inverted) conditional provides similar readability.

Comment (carterkozak) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCompletableFutureUsage.java]: 👍

**Pull request merged**